src/main/java/hudson/plugins/s3/cloudfront/CloudFrontInvalidatePublisher.java

```diff
@@ -0,0 +1,188 @@
+package hudson.plugins.s3.cloudfront;
+
+import hudson.Extension;
+import hudson.Launcher;
+import hudson.Util;
+import hudson.model.BuildListener;
+import hudson.model.Describable;
+import hudson.model.Result;
+import hudson.model.AbstractBuild;
+import hudson.model.AbstractProject;
+import hudson.plugins.s3.S3BucketPublisher;
+import hudson.plugins.s3.S3Profile;
+import hudson.tasks.BuildStepDescriptor;
+import hudson.tasks.BuildStepMonitor;
+import hudson.tasks.Publisher;
+import hudson.tasks.Recorder;
+import hudson.util.CopyOnWriteList;
+import hudson.util.ListBoxModel;
+
+import java.io.IOException;
+import java.io.PrintStream;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+
+import jenkins.model.Jenkins;
+
+import org.apache.commons.lang.StringUtils;
+import org.kohsuke.stapler.DataBoundConstructor;
+
+public final class CloudFrontInvalidatePublisher extends Recorder implements Describable<Publisher> {
+	
+	@Extension
+	public static final DescriptorImpl DESCRIPTOR = new DescriptorImpl();
+
+	private String profileName;
+	private final List<InvalidationEntry> invalidationEntries;
+
+	public CloudFrontInvalidatePublisher(){
+		super();
+		this.invalidationEntries = Collections.emptyList();
+	}
+	
+	@DataBoundConstructor
+	public CloudFrontInvalidatePublisher(String profileName, List<InvalidationEntry> invalidationEntries) {
+		if (StringUtils.isBlank(profileName)) {
+			// defaults to the first one
+			S3Profile[] sites = DESCRIPTOR.getProfiles();
+			if (sites.length > 0)
+				profileName = sites[0].getName();
+		}
+
+		this.profileName = profileName;
+		this.invalidationEntries = invalidationEntries;
+
+	}
+
+	public String getProfileName() {
+		return profileName;
+	}
+
+	public List<InvalidationEntry> getInvalidationEntries() {
+		return invalidationEntries;
+	}
+
+	public S3Profile getProfile() {
+		return getProfile(profileName);
+	}
+
+	public static S3Profile getProfile(String profileName) {
+		S3Profile[] profiles = DESCRIPTOR.getProfiles();
+
+		if (profileName == null && profiles.length > 0)
+			// default
+			return profiles[0];
+
+		for (S3Profile profile : profiles) {
+			if (profile.getName().equals(profileName))
+				return profile;
+		}
+		return null;
+	}
+
+	protected void log(final PrintStream logger, final String message) {
+		logger.println(StringUtils.defaultString(getDescriptor().getDisplayName()) + " " + message);
+	}
+
+	@Override
+	public boolean perform(AbstractBuild<?, ?> build, Launcher launcher, BuildListener listener) throws InterruptedException, IOException {
+
+		final boolean buildFailed = build.getResult() == Result.FAILURE;
```

Comment:
Each entry has checkbox whether to perform invalidation if build failed. It was done the same way as upload to s3, if resources were uploaded then probably they should be invalidated